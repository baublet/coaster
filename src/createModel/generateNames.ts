export function generateNames(
  name: string
): {
  singularCamel: string;
  singularPascal: string;
  pluralCamel: string;
  pluralPascal: string;
} {
  return {
    singularCamel: name,
    pluralCamel: name,
    pluralPascal: name,
    singularPascal: name,
  };
}
