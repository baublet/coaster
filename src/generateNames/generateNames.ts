export function generateNames(
  name: string
): {
  raw: string;
  rawCamel: string;
  rawPascal: string;
  singularCamel: string;
  singularPascal: string;
  pluralCamel: string;
  pluralPascal: string;
} {
  return {
    raw: name,
    rawCamel: name,
    rawPascal: name,
    singularCamel: name,
    pluralCamel: name,
    pluralPascal: name,
    singularPascal: name,
  };
}
