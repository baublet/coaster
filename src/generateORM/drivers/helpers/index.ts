export { shouldExclude } from "./shouldExclude";
export { dbTypeToTypeScriptType } from "./dbTypeToTypeScriptType";
