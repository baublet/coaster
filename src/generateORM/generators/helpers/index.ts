export { getName } from "./getName";
export { getSchemaAndTablePath } from "./getSchemaAndTablePath";
export { getTypeName } from "./getTypeName";
export { getDefaultJavascriptForColumnType } from "./getDefaultJavascriptForColumnType";
export { TemplateManager, getTemplateManager } from "./templateManager";
