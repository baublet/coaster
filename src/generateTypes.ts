import {
  Schema,
  SchemaNodeType,
  SchemaEntity,
  isSchemaEntityConfiguration
} from "./schema";

export interface GenerateTypesFromNodesArguments {
  entities: SchemaEntity[];
}

export interface GenerateTypesArguments {
  schema: Schema;
}

function schemaEntityToTypeDefinition(entity: SchemaEntity): string {
  const name = entity.names.pascal;

  const typeParts: string[] = [];
  typeParts.push(`interface ${name} {`);
  const properties = Object.keys(entity.nodes);

  for (const property of properties) {
    const node = entity.nodes[property];

    const propertyParts: string[] = [];
    propertyParts.push(`\t${property}`);

    const isNullable = isSchemaEntityConfiguration(node)
      ? Boolean(node.nullable)
      : false;
    if (isNullable) {
      propertyParts.push("?");
    }

    propertyParts.push(": ");

    const type = isSchemaEntityConfiguration(node) ? node.type : node;
    switch (type) {
      case SchemaNodeType.BOOLEAN:
        propertyParts.push("boolean");
        break;
      case SchemaNodeType.NUMBER:
        propertyParts.push("number");
        break;
      case SchemaNodeType.STRING:
        propertyParts.push("string");
        break;
      default:
        throw new Error(`Unknown schema node property type: ${type}`);
    }
    propertyParts.push(";");
    typeParts.push(propertyParts.join(""));
  }

  typeParts.push("}");

  return typeParts.join("\n");
}

export function generateTypes({ schema }: GenerateTypesArguments): string {
  const types: string[] = [];

  for (const schemaNode of schema.nodes) {
    types.push(schemaEntityToTypeDefinition(schemaNode));
  }

  return types.join("\n\n");
}
