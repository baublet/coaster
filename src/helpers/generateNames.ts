import { camelCase, pascalCase, snakeCase } from "change-case";
import pluralize from "pluralize";
import memoize from "lodash.memoize";

export interface GeneratedNames {
  /**
   * The canonical name. Not safe for use in databases or in JS.
   */
  canonical: string;
  /**
   * Plural version of canonical.
   */
  plural: string;
  /**
   * Often the same as above, but in cases where canonical needs to vary from
   * the canonical, use this.
   */
  original: string;
  /**
   * Pluralized version of original.
   */
  originalPlural: string;
  /**
   * A database- and javascript-safe version of the string. Suitable for using
   * as database column names.
   */
  safe: string;
  /**
   * The plural version of the safe name.
   */
  pluralSafe: string;
  /**
   * The pascal-case name
   */
  pascal: string;
  /**
   * The camel-case name
   */
  camel: string;
}

/**
 * Takes a string and returns a structure of various names for use within an
 * application. If `original` is passed in, we use that as the original value.
 * This is useful if the name passed in is already transformed in some way.
 * @param name
 * @param original
 */
function generateNames(
  name: string,
  original: string | false = false
): GeneratedNames {
  const safe = snakeCase(name);
  return {
    canonical: name,
    original: original || name,
    originalPlural: pluralize(original || name, 2),
    plural: pluralize(name, 2),
    pluralSafe: pluralize(safe, 2),
    safe: safe,
    pascal: pascalCase(name),
    camel: camelCase(name)
  };
}

export default memoize<typeof generateNames>(generateNames);
