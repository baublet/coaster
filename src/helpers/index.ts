export type Maybe<T> = T | void;
export { GeneratedNames, generateNames } from "./generateNames";
export { isEmptyObject } from "./isEmptyObject";
