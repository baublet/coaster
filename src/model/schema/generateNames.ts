import toSnakeCase from "snake-case";

export interface SchemaNodeNames {
  canonical: string;
  safe: string;
}

export default function generateNames(name: string): SchemaNodeNames {
  return {
    canonical: name,
    safe: toSnakeCase(name)
  };
}
