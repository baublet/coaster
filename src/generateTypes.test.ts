import { generateTypes } from "./generateTypes";
import { generateNames } from "helpers/generateNames";
import { SchemaNodeType } from "schema";

it("generates a type", () => {
  expect(
    generateTypes({
      schema: {
        name: "Test schema",
        description: "A schema for an application",
        nodes: [
          {
            names: generateNames("user"),
            nodes: {
              id: SchemaNodeType.NUMBER,
              name: SchemaNodeType.STRING,
              active: SchemaNodeType.BOOLEAN,
              jobTitle: {
                type: SchemaNodeType.STRING,
                nullable: true
              }
            }
          },
          {
            names: generateNames("post"),
            nodes: {
              id: SchemaNodeType.NUMBER,
              createdDate: SchemaNodeType.NUMBER,
              modifiedDate: SchemaNodeType.NUMBER,
              title: SchemaNodeType.STRING,
              content: SchemaNodeType.STRING
            }
          }
        ]
      }
    })
  ).toMatchSnapshot();
});
