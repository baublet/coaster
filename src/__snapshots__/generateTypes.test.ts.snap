// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates a type 1`] = `
"interface User {
	id: number;
	name: string;
	active: boolean;
	jobTitle?: string;
}

interface Post {
	id: number;
	createdDate: number;
	modifiedDate: number;
	title: string;
	content: string;
}"
`;
